Source/WebKit/WPEPlatform/wpe/WPECursorTheme.cpp

```diff
@@ -73,42 +73,54 @@ static GUniquePtr<char> cursorsPath(const char* basePath, Vector<GUniquePtr<char
+static std::unique_ptr<CursorTheme> tryCreateTheme(const char* basePath, uint32_t size)
 {
-    auto findThemePath = [](const char* name) -> GUniquePtr<char> {
-        GUniquePtr<char> path(g_build_filename(g_get_user_data_dir(), "icons", name, nullptr));
-        if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-            return path;
-
-        path.reset(g_build_filename(g_get_home_dir(), ".icons", name, nullptr));
-        if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-            return path;
-
-        auto* dataDirs = g_get_system_data_dirs();
-        for (unsigned i = 0; dataDirs[i]; ++i) {
-            path.reset(g_build_filename(dataDirs[i], "icons", name, nullptr));
-            if (g_file_test(path.get(), G_FILE_TEST_IS_DIR))
-                return path;
-        }
+    if (!g_file_test(basePath, G_FILE_TEST_IS_DIR))
         return nullptr;
-    };
 
-    if (auto basePath = findThemePath(name)) {
-        Vector<GUniquePtr<char>> inheritedThemes;
-        auto path = cursorsPath(basePath.get(), inheritedThemes);
-        if (path || !inheritedThemes.isEmpty()) {
-            if (!path) {
-                // If there's no cursors path, use the first inherited theme.
-                path = WTFMove(inheritedThemes[0]);
-                inheritedThemes.removeAt(0);
-            }
-
-            return makeUnique<CursorTheme>(WTFMove(path), size, WTFMove(inheritedThemes));
-        }
-        g_warning("Could not find any cursors (search started from '%s')", basePath.get());
-    } else
-        g_warning("Could not find any base paths for cursors search");
+    Vector<GUniquePtr<char>> inheritedThemes;
+    auto path = cursorsPath(basePath, inheritedThemes);
+    if (!path && inheritedThemes.isEmpty())
+        return nullptr;
+
+    if (!path) {
+        // If there's no cursors path, use the first inherited theme.
+        path = WTFMove(inheritedThemes[0]);
+        inheritedThemes.removeAt(0);
+    }
+
+    return makeUnique<CursorTheme>(WTFMove(path), size, WTFMove(inheritedThemes));
+}
+
+std::unique_ptr<CursorTheme> CursorTheme::create(const char* name, uint32_t size)
+{
+    GUniquePtr<char> path(g_build_filename(g_get_user_data_dir(), "icons", name, nullptr));
+    if (auto theme = tryCreateTheme(path.get(), size))
+        return theme;
+
+    path.reset(g_build_filename(g_get_home_dir(), ".icons", name, nullptr));
+    if (auto theme = tryCreateTheme(path.get(), size))
+        return theme;
+
+    auto* dataDirs = g_get_system_data_dirs();
+    for (unsigned i = 0; dataDirs[i]; ++i) {
+        path.reset(g_build_filename(dataDirs[i], "icons", name, nullptr));
+        if (auto theme = tryCreateTheme(path.get(), size))
+            return theme;
+    }
+
+    // If not the default, try with the default.
+    if (g_strcmp0(name, "default")) {
+        if (auto theme = CursorTheme::create("default", size))
+            return theme;
+    }
```

Comment:
I might be miss-reading this, but in an scenario where no cursors are present, we would attempt the Adwaita and print the `g_warning` below twice, wouldn't we?

In line 114, we would recursively call  back here attempting to create a `default` theme, it would go to line 113, skip, continue into fallback to Adwaita in line 118, fail, then print warning in 123, return `nullptr`. The caller would, in line 114, continue down as well, fallback again to Adwaita.

So we probably should not try _Adwaita_ if we're attempting to find the `default` theme.

```suggestion
    // If not the default, try with the default.
    if (g_strcmp0(name, "default")) {
        if (auto theme = CursorTheme::create("default", size))
            return theme;
    } else
        return nullptr
```

the above should be able too continue into Adwaita and print a single warning if `theme` is null in line 115.

Reply:
Right, good point!

Reply:
But we should return early when not the default, I'll update the patch.